src/BotwinExtensions.cs

```diff
@@ -44,41 +45,52 @@ public static IApplicationBuilder UseBotwin(this IApplicationBuilder builder, Bo
                 {
                     var moduleType = module.GetType();
 
-                    foreach (var route in module.Routes)
+                    foreach (var route in module.Routes.Keys)
                     {
-                        routeBuilder.MapVerb(route.verb, route.path, ctx =>
-                        {
-                            var statusCodeHandlers = ctx.RequestServices.GetServices<IStatusCodeHandler>();
-
-                            var requestScopedModule = ctx.RequestServices.GetRequiredService(moduleType) as BotwinModule;
-
-                            // TODO: Use the handler from the resolved 'requestScopedModule' instead of the one from the "startup scope".
-                            var routeHandler = requestScopedModule.Routes.FirstOrDefault(x => x.path == route.path && x.verb == route.verb).handler;
-                            var handler = CreateModuleBeforeAfterHandler(requestScopedModule, routeHandler);
-
-                            var finalHandler = CreateFinalHandler(handler, statusCodeHandlers);
-
-                            return finalHandler.Invoke(ctx);
-                        });
+                        routeBuilder.MapVerb(route.verb, route.path, CreateRouteHandler(route, moduleType));
                     }
                 }
             }
 
             return builder.UseRouter(routeBuilder.Build());
         }
 
-        private static RequestDelegate CreateFinalHandler(RequestDelegate handler, IEnumerable<IStatusCodeHandler> statusCodeHandlers)
+        private static RequestDelegate CreateRouteHandler((string verb, string path) route, Type moduleType)
         {
-            RequestDelegate finalHandler = async (ctx) =>
+            return async (HttpContext ctx) =>
             {
+                var module = ctx.RequestServices.GetRequiredService(moduleType) as BotwinModule;
+                if (!module.Routes.TryGetValue((route.verb, route.path), out var routeHandler))
+                    throw new InvalidOperationException($"Route {route.verb} '{route.path}' was no longer found");
+
+                // begin handling the request
                 if (HttpMethods.IsHead(ctx.Request.Method))
                 {
                     //Cannot read the default stream once WriteAsync has been called on it
                     ctx.Response.Body = new MemoryStream();
                 }
 
-                await handler(ctx);
+                // run the module handlers
+                bool shouldContinue = true;
+                if (module.Before != null)
+                {
+                    shouldContinue = await module.Before(ctx);
```

Comment:
I think this should return at this point if `shouldContinue` is false?

Reply:
I originally had this to return, however I realized that in the original implementation, the `return` only affected the “before and after handler”. So returning here would be wrong as it would terminate the full handler, including what happened inside the final handler and e.g. the status code handlers.

So returning `false` from the before handler will prevent the module’s route handler from running (and the after handler) but not terminate the response immediately to allow the other components from finishing their job.